[AArch64][NEON] Add famax/famin codegen patterns

SpencerAbson opened the pull request:
- min(abs(a), abs(b)) -> famin(a, b), max(abs(a), abs(b))-> famax(a, b)

- Changes to LLVM
    - llvm/lib/Target/AArch64InstrInfo.td
	    - Add pattern for NEON types
     - +llvm/test/CodeGen/AArch64/aarch64-neon-faminmax.ll
	    - Add tests with and without +faminmax flag.

SpencerAbson committed:
Add famax/famin codegen patterns for NEON

- Replace min(abs(a), abs(b)) with famin(a, b)
- Replace max(abs(a), abs(b)) with famax(a, b)

- llvm/lib/Target/AArch64InstrInfo.td
	- Add pattern for NEON types
- +llvm/test/CodeGen/AArch64/aarch64-neon-faminmax.ll
	- Add tests with and without +faminmax flag.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -10155,7 +10155,19 @@ let Predicates = [HasFP8] in {
  defm FAMAX : SIMDThreeSameVectorFP<0b0, 0b1, 0b011, "famax", null_frag>;
  defm FAMIN : SIMDThreeSameVectorFP<0b1, 0b1, 0b011, "famin", null_frag>;
-} // End let Predicates = [HasFAMAXMIN]
+} // End let Predicates = [HasFAMINMAX]
+
+let Predicates = [HasNEON, HasFAMINMAX] in {
```
You shouldn't need to break the block here but instead use the new `Predicates` definition for the instruction defs as well as your new isel patterns.

SpencerAbson replied:
Thanks, done.

paulwalker-arm commented on llvm/test/CodeGen/AArch64/aarch64-neon-faminmax.ll:
```diff
@@ -0,0 +1,207 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=aarch64 -mattr=+faminmax -verify-machineinstrs %s -o - | FileCheck %s
+; RUN: llc -mtriple=aarch64 -verify-machineinstrs %s -o - | FileCheck %s --check-prefix=CHECK-NO-FAMINMAX
+
```
There's no great consistency here but do you mind adding `target triple = "aarch64-unknown-linux-gnu"` here and then you can remove the `-mtriple` from the RUN lines.

paulwalker-arm replied:
I have and always will disagree with this.  I think RUN lines should be minimised to show the relevant part of the testing.  If a test depends on multiple triples then sure add them to the RUN line.  If, like it is for the majority of cases, the triple is not relevant then it doesn't belong on the RUN line.

davemgreen replied:
Sigh, I shouldn't have said anything. The more important part was about adding fminnum support. I don't believe this makes things simpler though, and we have several thousand tests that rightly put the triple in the run line. At the very least it keeps everything close together (triple and attibutes) which makes then easier to read, and in the end of the day being consistent is more important. We will always need triples in some run lines, so it's best to keep them in all. Thanks.

SpencerAbson committed:
Update prediction of NEON famin/famax

- Change prediction of NEON famin/fmax instructions to
 [HasNEON,HasFAMINMAX]
- Use target triple string in faminmax llc test instead
  of 'mtriple' flag.

SpencerAbson commented:
Thank you, Paul.

davemgreen reviewed (commented):
Hi - Matching from a fminimum (denormals aside) sounds like it should be correct, that sounds good. Unfortunately that does make it less useful. Did you consider adding patterns for fminnum with nnan too? They are much more likely to come up in practice if people are using fast-math.
